Jobs/Singlepoint.py

```diff
@@ -0,0 +1,151 @@
+import sys
+import subprocess
+import numpy as np
+
+#   Imports From Existing Libraries
+
+#   Imports Of Custom Libraries
+
+#   Imports From Custom Libraries
+from Logging.Logger import Logger
+from Logging.WriteOutput import Output
+from Jobs import QM_job, MM_job
+from Generators._helper import filter_xyzq, _flatten
+from Generators.GeneratorGeometries import read_gmx_structure_header, read_gmx_structure_atoms, read_gmx_box_vectors, write_g96
+from Generators.GeneratorEnergies import GeneratorEnergies, GeneratorForces
+
+#   // TODOS & NOTES //
+#   TODO:
+#   - Add logger
+#   NOTE:
+
+#   // CLASS & METHOD DEFINITIONS //
+class Singlepoint():
+
+    '''
+    This Class Performs A Singlepoint Calculation
+    '''
+
+    def __init__(self, dict_input_userparameters, class_system, class_topology, class_pcf, str_directory_base) -> None:
+        '''
+        ------------------------------ \\
+        EFFECT: \\
+        --------------- \\
+        XX\\
+        ------------------------------ \\
+        INPUT: \\
+        --------------- \\
+        input_dict: dict -> Dictionary Of The User Parameter Input \\
+        class_system: class -> Class Object Of The System \\
+        class_topology: class -> Class Object Of The Topology \\
+        class_pcf: class -> Class Object Of The Pointchargefield \\
+        str_directory_base: str -> Directory Path \\ 
+        ------------------------------ \\
+        RETURN: \\
+        --------------- \\
+        NONE \\
+        ------------------------------ \\
+        '''
+
+        self.dict_input_userparameters = dict_input_userparameters
+        self.system = class_system
+        self.class_topology_qmmm = class_topology
+        self.PCF = class_pcf
+        self.str_directory_base = str_directory_base
+
+        #   XX AJ check how to deal with nma flag later
+        self.nmaflag = 0
+
+        #   Initialize QM Class (XX AJ differentiate between qm program here?)
+        if self.dict_input_userparameters['qmcommand'] == 'g16':
+            self.class_qm_job = QM_job.QM_gaussian(self.dict_input_userparameters, self.system, self.class_topology_qmmm, self.PCF, self.str_directory_base)
+        elif self.dict_input_userparameters['qmcommand'] == 'orca':
+            pass
+
+        #   Initialize MM Class
+        self.class_mm_job = MM_job.MM(self.dict_input_userparameters, self.system, self.class_topology_qmmm, self.PCF, self.str_directory_base)
+
+        #   Initialize QMMM Energy And Forces Generator Classes
+        self.class_qmmm_energy = GeneratorEnergies(self.dict_input_userparameters, self.system, self.class_topology_qmmm, self.PCF, self.str_directory_base, self.class_qm_job, self.class_mm_job)
+        self.class_qmmm_forces = GeneratorForces(self.dict_input_userparameters, self.system, self.class_topology_qmmm, self.PCF, self.str_directory_base, self.class_qm_job, self.class_mm_job)
+
+
+        #   Run (Initial) Singlepoint Calculation
+        self.run_calculation()
```

Comment:
Are you sure you want to run the actual calculation already in `__init__`? Makes me wonder a bit why this is a class in the first place instead of just a function.

Reply:
For singlepoint calculations you're correct. For optimizations we're calling the run_calculation function more than once, using the initial instance of that class.

Reply:
Ok, but in this way the use of the `Singlepoint` class is reduced to one-off usage like

```python
Singlepoint(*args)
# Init, calculation work, and output production all in one go
# The instance can be thrown away afterwards
```

and in this case, this is just the same as executing a big function with side-effects. As far as I can see it right now, your other classes (`Optimization`, `Output`, etc) have a similar behaviour.

You lose the actual advantage of having classes here. I would expect a usage pattern of something more like:

```python
singlepoint = Singlepoint(*args)  # Only init
singlepoint.run()  # Only calculation
singlepoint.output  # Inspect output
singlepoint.write()   # Write output 
```

This gives you the flexibility to create and use the functionality encapsulated by the class in more different situations. The class is not just a namespace to collect non-controllable subfunctions but really a handy standalone object that bundles data with functionality and can be passed around, re-used, pickled and whatnot. It also decouples the individual bits and exposes them in a meaningful way. You can still have a high-level user function in the end that makes use of the class in a particular way.

Let's maybe talk about it in a meeting, so I can better understand your motivation for making these classes.